src/components/select/filterable-select/filterable-select.component.js

```diff
@@ -0,0 +1,293 @@
+import React, {
+  useState,
+  useEffect,
+  useRef,
+  useCallback
+} from 'react';
+import PropTypes from 'prop-types';
+import invariant from 'invariant';
+import SelectTextbox from '../select-textbox';
+import guid from '../../../utils/helpers/guid';
+import Events from '../../../utils/helpers/events';
+import withFilter from '../utils/with-filter.hoc';
+import SelectList from '../select-list';
+import { formInputPropTypes } from '../select-textbox/select-textbox.component';
+
+const FilterableSelect = ({
+  value,
+  defaultValue,
+  id,
+  name,
+  children,
+  inputRef,
+  onOpen,
+  onChange,
+  onClick,
+  onFocus,
+  ...textboxProps
+}) => {
+  const selectListId = guid();
+  const containerRef = useRef();
+  const listboxRef = useRef();
+  const isControlled = useRef(value !== undefined);
+  const [textboxRef, setTextboxRef] = useState();
+  const [isOpen, setOpenState] = useState(false);
+  const [textValue, setTextValue] = useState('');
+  const [selectedValue, setSelectedValue] = useState('');
+  const [highlightedValue, setHighlightedValue] = useState('');
+  const [filterText, setFilterText] = useState('');
+
+  const assignInput = useCallback((input) => {
```

Comment:
1. I agree, I'll implement a forward ref.
2. I need to get the parent element of the input, to do it I need it to be set in the state. When you have your method wrapped in `useCallback` hook it will always reference proper htmlElement as it's happening here, so I can't see any issues with that. This is a ref to the main element of the component and I can't see a case when it's replaced with another one during the life of that component.

In their docs they are saying that the ref shouldn't be defined as an inline function, and it's not an inline function so the mentioned caveat does not concerns us here.

Reply:
1. ✅ 
2. In the past it's been when a user has tried to access a ref within a callback, e.g. accessing the ref in "onBlur". I haven't been able to reproduce the example, so it looks like we're fine 😄 
We do still however need to handle callback refs, because we're setting the ref manually, we don't get that for free.